src/ncopt/sqpgs/main.py

```diff
@@ -449,169 +445,68 @@ def __init__(self, dim, p0, pI, pE, assert_tol):
-        self.update() before solving in order to have the correct subproblem data.
-
-        self.d: array
-            search direction
-
-        self.lambda_f: array
-            KKT multipier for objective.
-
-        self.lambda_gE: list
-            KKT multipier for equality constraints.
-
-        self.lambda_gI: list
-            KKT multipier for inequality constraints.
-
-        """
-        cx.solvers.options["show_progress"] = False
-
-        iG = np.vstack((self.inG, self.nonnegG))
-        ih = np.hstack((self.inh, self.nonnegh))
-
-        qp = cx.solvers.qp(
-            P=cx.matrix(self.P), q=cx.matrix(self.q), G=cx.matrix(iG), h=cx.matrix(ih)
-        )
-
-        self.status = qp["status"]
-        self.cvx_sol_x = np.array(qp["x"]).squeeze()
-
-        self.d = self.cvx_sol_x[: self.dim]
-        self.z = self.cvx_sol_x[self.dim]
-
-        self.rI = self.cvx_sol_x[self.dim + 1 : self.dim + 1 + self.nI]
-        self.rE = self.cvx_sol_x[self.dim + 1 + self.nI :]
-
-        assert len(self.rE) == self.nE
-        assert np.all(
-            self.rI >= -self.assert_tol
-        ), f"Array should be non-negative, but minimal value is {np.min(self.rI)}."
-        assert np.all(
-            self.rE >= -self.assert_tol
-        ), f"Array should be non-negative, but minimal value is {np.min(self.rE)}."
-
-        # extract dual variables = KKT multipliers
-        self.cvx_sol_z = np.array(qp["z"]).squeeze()
-        lambda_f = self.cvx_sol_z[: self.p0 + 1]
-
-        lambda_gI = list()
-        for j in np.arange(self.nI):
-            start_ix = self.p0 + 1 + (1 + self.pI)[:j].sum()
-            lambda_gI.append(self.cvx_sol_z[start_ix : start_ix + 1 + self.pI[j]])
-
-        lambda_gE = list()
-        for j in np.arange(self.nE):
-            start_ix = self.p0 + 1 + (1 + self.pI).sum() + (1 + self.pE)[:j].sum()
-
-            # from ineq with +
-            vec1 = self.cvx_sol_z[start_ix : start_ix + 1 + self.pE[j]]
+        self.d = cp.Variable(self.dim)
+        self._problem = None
 
-            # from ineq with -
-            vec2 = self.cvx_sol_z[
-                start_ix + (1 + self.pE).sum() : start_ix + (1 + self.pE).sum() + 1 + self.pE[j]
-            ]
+    @property
+    def nI(self) -> int:
+        return len(self.pI)
 
-            # see Direction.m line 620 in the original Matlab code
-            lambda_gE.append(vec1 - vec2)
+    @property
+    def nE(self) -> int:
+        return len(self.pE)
 
-        self.lambda_f = lambda_f.copy()
-        self.lambda_gI = lambda_gI.copy()
-        self.lambda_gE = lambda_gE.copy()
+    @property
+    def has_ineq_constraints(self) -> bool:
+        return self.nI > 0
 
-        return
+    @property
+    def has_eq_constraints(self) -> bool:
+        return self.nE > 0
 
-    def initialize(self):
-        """
-        The quadratic subrpoblem we solve in every iteration is of the form:
+    @property
+    def problem(self) -> cp.Problem:
```

Comment:
What is the ruleset on whether to define a property or to initialize in the `__init__`? To me this looks slightly overcomplicated...